### DISCUSSION TITLE
API Portal returns generic error instead of redirecting to web login

### DISCUSSION THREAD

**Discussion by gbarbon** [mia-platform/community#48] [Q&A] [answered]
The API Portal is returning a generic “Something went wrong“ (see the image below) error instead of redirecting to the login page when the endpoints are protected and the user is not logged in.

![image](https://user-images.githubusercontent.com/6908044/194587119-0d1e00fb-8147-4918-a94e-3d6242ce65ea.png)

This is due to the `https://[redacted part]/documentations/api-portal/api/openapi/v3/json`  answering with a 401 error. I would expect the login page to be loaded instead.

When logged in, the API Portal works correctly. I suspect this is a configuration issue, but I don't know what I'm missing.

**Comment by fredmaggiowski** [comment]
Hi, how can the API Portal redirect you to a login page? It does not know whether you are using a login page or not or where it is served 🤔

**Comment by davidebianchi** [comment]
Hi @gbarbon, if you want to expose the API Portal to invoke API under authentication, you should setup it under a login page before the portal. Or leave the user insert the authentication headers.

**Comment by gbarbon** [comment, marked as answer]
You're both right. The description of the problem I have written is a bit misleading. This is not actually an API Portal issue, but it's more related to the API Gateway I guess. The redirect should be carried by the API Gateway, but it's not happening as intended. Just to give some more context, in the same project there is the Backoffice that has the redirect correctly working when the user is not authenticated.

**Reply by fredmaggiowski** [reply]
The redirect you're referring to happens when you try to access the backoffice itself or even when the backoffice is fetching some API?

**Reply by gbarbon** [reply]
I'm referring to the redirect of the Backoffice itself. When I call the Backoffice endpoint, the response headers contain the attribute `location: [redacted-part]/web-login?redirect=/`

**Reply by davidebianchi** [reply]
You should configure the API Gateway to do the correct redirect on 401.

From the advanced section of the console, for api-gateway, in root location extension, put this:
```
error_page 401 = /error401.$extension_by_http_accept;
```

and in server extension
```
location /error401.html {
  include /etc/nginx/customization.d/header-debug.conf;

  return 302 '$original_request_scheme://$original_request_host/web-login?appId=cms&redirect=$original_request_uri$is_args$args';
}

location /error401.json {
  include /etc/nginx/customization.d/header-debug.conf;

  return 401 $content_401;
}
```

With this configuration, the error code 401 will be redirect to the path, which responds with the redirect.

**Reply by gbarbon** [reply]
We already have something similar in the server extension in the API Gateway advanced settings:

```
error_page 401 = @error401;

location @error401 {
  include /etc/nginx/customization.d/header-debug.conf;

  if ($type = "text/html") {
    add_header Set-Cookie "mia_client_key=[redacted]; path=/; secure; HttpOnly;";
    return 302 '$original_request_scheme://$original_request_host/web-login?redirect=$original_request_uri$is_args$args';
  }

  default_type $type;
  return 401 $content_401;
}
```

I see that it differs from the code you suggest. Shouldn't this provide the same result as your version?

As for the root location extension, I've just updated it but it didn't change the result.

**Reply by gbarbon** [reply]
Btw I also tried your configuration, and the api-gateway container is restarting with the following error:
```
2022/10/12 07:29:00 [emerg] 1#1: duplicate location "/error401.html" in /etc/nginx/customization.d/default-error-page.conf:21
nginx: [emerg] duplicate location "/error401.html" in /etc/nginx/customization.d/default-error-page.conf:21
```

**Reply by fredmaggiowski** [reply]
The location should already be available on the main nginx upstream so you might not need to specify it

**Reply by gbarbon** [reply]
@fredmaggiowski you mean this part `location /error401.html`? What should I write instead? I think that I still need the part where the `302` is returned.

**Reply by gbarbon** [reply]
I've checked with @fredmaggiowski , there was an error in my configuration: in the `GET /documentations/api-portal/` route of the `/documentations/api-portal` endpoint, the "All routes access security" check was not marked as inherited (an the authentication was not required). Because of that, the API Portal was not redirecting the user to the login page, believing the user as authenticated, even if he was not. Thanks @fredmaggiowski !

**Reply by fredmaggiowski** [reply]
You're welcome
